### DISCUSSION TITLE
Design of IPeer

### DISCUSSION THREAD

**Discussion by tsuckow** [NethermindEth/dotnet-libp2p#81] [Q&A]
I noticed the implementation of IPeer only can hold a single multiaddress. This could make sense if it was for a specific connection with a peer, but I don't think that is how this was intended.

Would it make sense to update this to handle multiple addresses?

```
/dnsaddr/am6.bootstrap.libp2p.io/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/dns4/am6.bootstrap.libp2p.io/tcp/443/wss/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/dns6/am6.bootstrap.libp2p.io/tcp/443/wss/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/ip4/147.75.87.27/tcp/4001/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/ip4/147.75.87.27/udp/4001/quic-v1/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/ip4/147.75.87.27/udp/4001/quic/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/ip6/2604:1380:4602:5c00::3/tcp/4001/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/ip6/2604:1380:4602:5c00::3/udp/4001/quic-v1/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/ip6/2604:1380:4602:5c00::3/udp/4001/quic/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
```

I'd really love to get DHT / Swarm working in this library.

But it also seems like if going that route, it would probably make sense to separate the connected status from IRemotePeer and build out classes representing a connection. As well as I don't think multiplexing has been completely worked out, but I could be wrong, I'm still reading through the code.

**Comment by tsuckow** [comment]
The RemotePeer implementation is NUTS. Instantiated before a connection is even made?!?!?

**Comment by flcl42** [comment]
> I noticed the implementation of IPeer only can hold a single multiaddress. This could make sense if it was for a specific connection with a peer, but I don't think that is how this was intended.
>
> Would it make sense to update this to handle multiple addresses?

@tsuckow hi! Yeah, I'm remaking PeerFactory into Swarm right now. We differently need to support multiple listening addresses and peer registry also. It will also make it easier to implement identify/push, make gossipsub more resilient 

> As well as I don't think multiplexing has been completely worked out, but I could be wrong, I'm still reading through the code.

Yamux update has been merged recently, it's way more complete now. What it lacks is window optimization on our side, we only handle what remote sends.

> The RemotePeer implementation is NUTS. Instantiated before a connection is even made?!?!?

Could you elaborate? Task result of `Task<IRemotePeer> DialAsync` should complete when connection is established and you can now dial with specific protocol.

**Reply by tsuckow** [reply]
ListenAsync in peerfactory creates an instance of remote peer with no identity or address then links it to peercontext. The way that handles new connections is some gymnastics with forking the remote peer. RemotePeer is treated more like a connection, but even that feels shady. We really need two different classes to represent what we know about a peer and another for the connection/subdial to the peer.

I started dabbling with refactoring IPeer to represent what we know about a peer (and jog my brain about thread safe collections in C#).

**Reply by flcl42** [reply]
Hm, if you mean `ListenAsync` from `ILocalPeer`, it provides `IListener` which is different from `IRemotePeer`, but I guess I see why it all causes confusion. There is no good source of peer information, and PeerContext is not a good source, it was made to cover it outside of **protocols** only.

**Reply by tsuckow** [reply]
I was referring to https://github.com/NethermindEth/dotnet-libp2p/blob/79237fba53601acda95362a8d08dc00bdeb1c34c/src/libp2p/Libp2p.Core/PeerFactory.cs#L73

Which creates an instance but at that moment, nothing is even connected yet.

It's growing pains, we'll get this figured out.

**Reply by tsuckow** [reply]
I'll have to contemplate the meaning of the universe. Ran into issues around the QUIC and TCP protocols since they are "special" in that they care about the remote peer address and only use the downstream connection as a cancelation token.

I really want to unravel the chicken and egg problem creating a peer node. Ideally, given an address and optional identity dialing could proceed without a peer object and once the remote identity was validated we would use the PeerFactory to make a RemotePeer + RemotePeerConnection. If the optionally provided identity didn't match, "dialing" would fail but the connection may be preserved.

Anyway, this makes me think a more complex set of interfaces around IProtocol might be nice, but what that looks like without being a PITA is eluding me.

My WIP changes FWIW https://github.com/tsuckow/dotnet-libp2p/commits/feature/multiple_addresses/

**Reply by flcl42** [reply]
> RemotePeer remotePeer = new(this, peer, peerContext); 

This part is hidden under the hood via private modifiers, seems OK to have incomplete intermediate state for it.

> Anyway, this makes me think a more complex set of interfaces around IProtocol might be nice, but what that looks like without being a PITA is eluding me.

Yeah, imho this is the way. Current hierarchy is more about single transport connection + several multiplexer streams. We can add swarm, wrapping the factory and other stuff around, make it working and then optimize hierarchy. Another thing I'm trying to add is event bus, which may help to simplify connection management a lot. Need time

Kad DHT can be added even now, with single address by the way.

**Reply by tsuckow** [reply]
At this point I'm just kinda going down my own path of learning so feel free to call me crazy and do your own thing. Or I'll get this to work and we can see what is worth keeping.

I started restructuring ProtocolStack to be more ProtocolRegistry and have it understand certain important kinds of protocols. I'm hoping to leverage dependancy injection a little more for initializing the registry, possibly moving the configuration of the listening for transports into a options object during injection.

```
        public IEnumerable<ITransport> Transports { get; }

        public IEnumerable<IMuxerProtocol> MuxerProtocols { get; }

        public IEnumerable<ISecurityProtocol> SecurityProtocols { get; }

        public IEnumerable<IIdentityProtocol> IdentityProtocols { get; }

        public IEnumerable<IProtocol> OtherProtocols { get; }
```

I feel a bit dirty making such radical changes, but as you said it was built for single connection so maybe it is not unexpected to need this to handle more complex scenarios

**Reply by flcl42** [reply]
> it was built for single connection

I mean, it's by purpose, we will always support single connection dialing, it's an essential API layer. We just need to add a wrapper for it that will manage connections.

**Comment by tsuckow** [comment]
Another day, another set of thoughts.

The implementation of IChannelFactory is interesting and I think overcomplicated. Once a multiplexor has been established on a connection it is possible for both sides to dial each other using it with a common interface.

**Reply by flcl42** [reply]
Yeah, there are two steps. Establishing kind of session, when multiplexer is ready to handle communication that can be initiated from any side; and then you can use it to communicate via any application level protocol. It would be nice to have a registry of established sessions (dictionary peerid = session), so you can:
- dial with application protocol, passing peerid instead of multiaddr;
- prevent from having several sessions with the same peer
- close all sessions at once
- react to session disconnection
- disconnect peer by peerid

Which sounds like swarm.

Another potential thing is peer store: a cache dictionary of peerid = multiaddr[], so in case you want to connect again, you can do it by peerid when possible.

**Reply by tsuckow** [reply]
I'm not convinced a wrapper would work very well without some fundamental changes.

We'll see ¯\\\_(ツ)\_/¯

**Comment by flcl42** [comment]
Multiple addresses are now supported

**Comment by Xyncgas** [comment]
Thanks for it I am seeing MultiAddress array in everywhere and it's important and I am grateful
May I change the codes so instead of using array it can use IEnumerable<MultiAddress>
Because I am seeing ObservableCollections<MultiAddress> for ISessionContex.(Peer :ILocalPeer).ListenAddresses and I am wanting to reduce allocations because p2p applications are connecting with a lot computers and if arrays are created for every ILocalPeer.DialAsync it's worrying me since my ram prices are getting higher and higher in 2025

**Reply by flcl42** [reply]
Yeah, let try to use IEnumerable. ListenAddresses btw is about what addresses the server is listening too, not remote peer's addresses or so. As a separate issue, ObservableCollection usage is not perfect probably, feel free to suggest better idea.
